include/pybind11/detail/argument_vector.h

```diff
@@ -201,32 +202,33 @@ struct argument_vector {
         using heap_vector = typename repr_type::heap_vector;
         heap_vector hv;
         hv.vec.reserve(reserved_size);
-        std::copy(ha.arr.begin(), ha.arr.begin() + ha.size, std::back_inserter(hv.vec));
+        static_assert(std::is_nothrow_move_constructible<T>::value,
+                      "this conversion is not exception safe");
+        static_assert(std::is_nothrow_move_constructible<heap_vector>::value,
+                      "this conversion is not exception safe");
```

Comment:
Slight nit: there is a less safe inplacement assignment path here if the objects are only std::is_nothrow_move_assignable. That's probably more complicated than it's worth.